### PR TITLE
{ARO} New ARO command_module for v2020_04_30 GA API to mange Azure Redhat OpenShit

### PR DESCRIPTION
**Pull request by mjudeikis**
**Description of PR (Mandatory)**  
Add ARO command_modules to azure CLI core 

**Testing Guide**  
Recoded API calls

**History Notes:**  
[ARO] az aro:  Add Azure RedHat OpenShift V4 aro command module  

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
add to S168

**Comment by mjudeikis** [comment]
@yonzhan thanks. Dates are fine for us. 

WIll see how we can fix missing peaces until this time. Let me know what is missing

**Comment by haroldrandom** [comment]
Hi, @mjudeikis Could you share some contexts?

**Comment by mjudeikis** [comment]
Hey @haroldrandom,

This is a new service (Azure RedHat OpenShift V4 (aka. ARO) developed together with RedHat and Microsoft. It is planned to GA in a month and a bit.
Python SDK is still being published (hence vendored SDK)
Code was generated using `azdev` scaffolding. 

At the moment this very basic command only.

File: .github/CODEOWNERS [hunk 1]
```diff
@@ -39,3 +39,4 @@
 /src/azure-cli/azure/cli/command_modules/servicebus/ @v-ajnava
 /src/azure-cli/azure/cli/command_modules/apim/ @kevinhillinger @jonlester
 /src/azure-cli/azure/cli/command_modules/iot/ @digimaun
+/src/azure-cli/azure/cli/command_modules/aro/ @mjudeiki @jim-minter
```
Comment (jim-minter) [inline .github/CODEOWNERS]: @mjudeikis?

File: src/azure-cli/azure/cli/command_modules/aro/custom.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+    if client_id is None:
+        app, client_secret = aad.create_application('aro-%s' % random_id)
+        client_id = app.app_id
+
+    client_sp = aad.get_service_principal(client_id)
+    if not client_sp:
+        client_sp = aad.create_service_principal(client_id)
+
+    rp_client_id = FP_CLIENT_ID
+    if rp_mode_development():
```
Comment (jim-minter) [inline src/azure-cli/azure/cli/command_modules/aro/custom.py]: I think we don't want this upstream

File: src/azure-cli/azure/cli/command_modules/aro/_rbac.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+from azure.cli.core.commands.client_factory import get_mgmt_service_client
+from azure.cli.core.commands.client_factory import get_subscription_id
+from azure.cli.core.profiles import get_sdk
+from azure.cli.core.profiles import ResourceType
+from msrestazure.tools import resource_id
+
+
+CONTRIBUTOR = 'b24988ac-6180-42a0-ab88-20f7382dd24c'
+DEVELOPMENT_CONTRIBUTOR = 'f3fe7bc1-0ef9-4681-a68c-c1fa285d6128'
```
Comment (jim-minter) [inline src/azure-cli/azure/cli/command_modules/aro/_rbac.py]: or this

File: .github/CODEOWNERS (hunk 1 above)
Comment (jim-minter) [reply .github/CODEOWNERS]: ie missing s?

File: src/azure-cli/HISTORY.rst [hunk 4]
```diff
@@ -43,6 +43,11 @@ Release History
 * az deployment {group/mg/sub/tenant} list: Support provisioningState filtering
 * az deployment: Fix the parse bug for comment under the last argument
 
+**ARO**
+
+* Added new command 'aro' for Azure RedHat OpenShift v4
+* Added support create, delete, list, show, update, list-credentials commands
+
```
Comment (haroldrandom) [inline src/azure-cli/HISTORY.rst:50]: Don't have to list history in this file, besides this file is not mean to modify by developer itself.

The content of this file comes from every PR title and PR body since last release. 

Please take a look this [doc](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests)

File: src/azure-cli/azure/cli/command_modules/aro/_params.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+                   tags_type)
+
+        c.argument('pull_secret',
+                   help='Pull secret of cluster.',
+                   validator=validate_pull_secret)
+        c.argument('domain',
+                   help='Domain of cluster.',
+                   validator=validate_domain)
+        c.argument('cluster_resource_group',
+                   help='Resource group of cluster.',
```
Comment (haroldrandom) [inline src/azure-cli/azure/cli/command_modules/aro/_params.py:41]: Why there is anohter resource group for cluster? Is it the same as the gloabl defaule resource group?

File: src/azure-cli/azure/cli/command_modules/aro/_params.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+        c.argument('apiserver_visibility',
+                   help='API server visibility.',
+                   validator=validate_visibility('apiserver_visibility'))
+
+        c.argument('ingress_visibility',
+                   help='Ingress visibility.',
+                   validator=validate_visibility('ingress_visibility'))
+
+        c.argument('vnet_resource_group_name',
```
Comment (haroldrandom) [inline src/azure-cli/azure/cli/command_modules/aro/_params.py:78]: Also confused like `cluster_resource_group`

File: src/azure-cli/azure/cli/command_modules/aro/_rbac.py (hunk 3 above)
Comment (haroldrandom) [inline src/azure-cli/azure/cli/command_modules/aro/_rbac.py:17]: Is this secret available related to some Azure resoruce?

File: src/azure-cli/azure/cli/command_modules/aro/_params.py [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+                   validator=validate_cidr('service_cidr'))
+
+        c.argument('master_vm_size',
+                   help='Size of master VMs.')
+
+        c.argument('worker_vm_size',
+                   help='Size of worker VMs.')
+        c.argument('worker_vm_disk_size_gb',
+                   help='Disk size in GB of worker VMs.',
+                   validator=validate_worker_vm_disk_size_gb)
```
Comment (haroldrandom) [inline src/azure-cli/azure/cli/command_modules/aro/_params.py:65]: You can add `type=int` to require the parameter value is an integer to save the work of validate function `_validate_int`

File: src/azure-cli/azure/cli/command_modules/aro/_params.py (hunk 5 above)
Comment (mjudeikis) [reply src/azure-cli/azure/cli/command_modules/aro/_params.py:41]: It is not. `resource_group` will contain cluster objects of its own , where `cluster_resource_group` is where the cluster resource components are deployed (if not provided we will create one for the user, but the user can override its name using this parameter and it will be locked for editing). This will be covered by user facing documentation, and it is the nature of the service architecture.

File: src/azure-cli/azure/cli/command_modules/aro/_params.py (hunk 7 above)
Comment (mjudeikis) [reply src/azure-cli/azure/cli/command_modules/aro/_params.py:65]: Will add this as follow-up issue[1] to our codebase, as all this code is a copy of our extension we use for PrivatePreview and development (https://github.com/Azure/ARO-RP/tree/master/python/az/aro). If this is ok, will fix in follow-up PR.

1. https://github.com/Azure/ARO-RP/issues/368

File: src/azure-cli/azure/cli/command_modules/aro/_params.py (hunk 6 above)
Comment (mjudeikis) [reply src/azure-cli/azure/cli/command_modules/aro/_params.py:78]: Service nature is "Bring your own VNET only". This value defines, where this VNET resides.

File: src/azure-cli/azure/cli/command_modules/aro/_rbac.py (hunk 3 above)
Comment (mjudeikis) [reply src/azure-cli/azure/cli/command_modules/aro/_rbac.py:17]: `CONTRIBUTOR` is Principal, assigned to this RP/Service, and is static across all the Azure. It is used for RP to get access into customer "BYO-VNET" from `vnet_resource_group_name`.

**Comment by mjudeikis** [comment]
> Generally okay except some question and please revert the HISTORY.rst.
> 
> Besides, could you record some tests?

Addresses the comments. Can test be added as follow-up PR not to block this one? 
We tested all this in our E2E (not python based) using extensions as this is not new code. But we will add some more testing here too.

File: src/azure-cli/azure/cli/command_modules/aro/_rbac.py (hunk 3 above)
Comment (haroldrandom) [reply src/azure-cli/azure/cli/command_modules/aro/_rbac.py]: Is there any risk that user could have change to manipulate something maliciously by using this?

**Comment by yonzhan** [comment]
@mjudeikis The commands in this PR is already in CLI extension, right? Could you please let us know the reason to move to CLI main repo?  We highly suggest that you still put the commands in CLI extension since we have plan to loose weight CLI main repo and CLI extension is more flexible to release new features. Meanwhile, for new RPs, we already deliver it in CLI extension.

File: src/azure-cli/azure/cli/command_modules/aro/_rbac.py (hunk 3 above)
Comment (mjudeikis) [reply src/azure-cli/azure/cli/command_modules/aro/_rbac.py]: This is Azure built in role. These should not change. https://docs.microsoft.com/en-us/azure/role-based-access-control/built-in-roles#general . As per my understanding, these should be static in AAD. Is this not a case?

```
{
    "assignableScopes": [
      "/"
    ],
    "description": "Lets you manage everything except access to resources.",
    "id": "/subscriptions/225e02bc-43d0-43d1-a01a-17e584a4ef69/providers/Microsoft.Authorization/roleDefinitions/b24988ac-6180-42a0-ab88-20f7382dd24c",
    "name": "b24988ac-6180-42a0-ab88-20f7382dd24c",
    "permissions": [
      {
        "actions": [
          "*"
        ],
        "dataActions": [],
        "notActions": [
          "Microsoft.Authorization/*/Delete",
          "Microsoft.Authorization/*/Write",
          "Microsoft.Authorization/elevateAccess/Action",
          "Microsoft.Blueprint/blueprintAssignments/write",
          "Microsoft.Blueprint/blueprintAssignments/delete"
        ],
        "notDataActions": []
      }
    ],
    "roleName": "Contributor",
    "roleType": "BuiltInRole",
    "type": "Microsoft.Authorization/roleDefinitions"
  },

```
P.S ACS, APPService, IOT and few other modules are taking the same aproach.

**Comment by mjudeikis** [comment]
> @mjudeikis The commands in this PR is already in CLI extension, right? Could you please let us know the reason to move to CLI main repo? We highly suggest that you still put the commands in CLI extension since we have plan to loose weight CLI main repo and CLI extension is more flexible to release new features. Meanwhile, for new RPs, we already deliver it in CLI extension.

@yonzhan extension is not published via extension repository. We publish it via our own link[1], and it is being used only by us, our integrated testing pipelines, and few PrivatePreview advanced customers.

Our goal for GA always was to land this in the core, so cloud shell would work without needing any changes. In addition, we have customer requirements to land this in the core due to internal security requirements for them. 

As you might be noticed this is not standard Azure service. ARO is jointly engineered with Microsoft and RedHat and is first of its kind (2nd Iteration of the service to be specific)

1. `az extension update -n aro --index https://az.aroapp.io/preview`

File: src/azure-cli/azure/cli/command_modules/aro/_rbac.py (hunk 3 above)
Comment (haroldrandom) [reply src/azure-cli/azure/cli/command_modules/aro/_rbac.py]: As for as I know, there is no hard code this string in CLI:

image

Could you get this string through graphrbac SDK like `az role`?

**Comment by mjudeikis** [comment]
@haroldrandom ready for the second round of review

File: src/azure-cli/azure/cli/command_modules/aro/_params.py [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+from azure.cli.core.commands.parameters import name_type
+from azure.cli.core.commands.parameters import resource_group_name_type
+from azure.cli.core.commands.parameters import tags_type
+from azure.cli.core.commands.validators import get_default_location_from_resource_group
+
+
+def load_arguments(self, _):
+    with self.argument_context('aro') as c:
+        c.argument('location',
+                   validator=get_default_location_from_resource_group)
```
Comment (haroldrandom) [inline src/azure-cli/azure/cli/command_modules/aro/_params.py:27]:
```suggestion
                   get_location_type(self.cli_ctx),
                   validator=get_default_location_from_resource_group),
```

File: src/azure-cli/azure/cli/command_modules/aro/custom.py [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+
+import azure.mgmt.redhatopenshift.models as v2020_04_30
+from azure.cli.command_modules.aro._aad import AADManager
+from azure.cli.command_modules.aro._rbac import assign_contributor_to_vnet
+from azure.cli.command_modules.aro._validators import validate_subnets
+from azure.cli.core.commands.client_factory import get_subscription_id
+from azure.cli.core.util import sdk_no_wait
+
+
+FP_CLIENT_ID = 'f1dd0a37-89c6-4e07-bcd1-ffd3d43d8875'
```
Comment (haroldrandom) [inline src/azure-cli/azure/cli/command_modules/aro/custom.py:17]: Why it's a staic hard code client ID? Will is cause security issue?

File: src/azure-cli/azure/cli/command_modules/aro/commands.py [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+
+def load_command_table(self, _):
+    aro_sdk = CliCommandType(
+        operations_tmpl='azure.cli.command_modules.aro.vendored_sdks.azure.mgmt.redhatopenshift.v2020_04_30.operations#OpenShiftClustersOperations.{}',  # pylint: disable=line-too-long
+        client_factory=cf_aro)
+
+    with self.command_group('aro', aro_sdk, client_factory=cf_aro, is_preview=True) as g:
+        g.custom_command('create', 'aro_create', supports_no_wait=True)
+        g.custom_command('delete', 'aro_delete', supports_no_wait=True, confirmation=True)
+        g.custom_command('list', 'aro_list', table_transformer=aro_list_table_format)
```
Comment (haroldrandom) [inline src/azure-cli/azure/cli/command_modules/aro/commands.py:21]: If ARO supports wait option, we highly recommend you provide command `wait` to give user the ability to wait until a desired operation is fine.
Please refer to https://github.com/Azure/azure-cli/blob/dev/doc/authoring_command_modules/authoring_commands.md

File: src/azure-cli/azure/cli/command_modules/aro/custom.py (hunk 9 above)
Comment (mjudeikis) [reply src/azure-cli/azure/cli/command_modules/aro/custom.py:17]: No, it will no. This is Azure FirstParty ID. It is given only for services it is guaranteed to be same and same across all clouds.

File: src/azure-cli/azure/cli/command_modules/aro/commands.py (hunk 10 above)
Comment (mjudeikis) [reply src/azure-cli/azure/cli/command_modules/aro/commands.py:21]: Added

File: src/azure-cli/azure/cli/command_modules/aro/_params.py (hunk 8 above)
Comment (mjudeikis) [reply src/azure-cli/azure/cli/command_modules/aro/_params.py:29]: added

**Comment by mjudeikis** [comment]
We are not quite sure where these CI failures are coming from. Code was not changed in any significant way but this started to fail n numerous way. We assume this might be that CI is not isolated from the overall az CLI ecosystem and due to some breaking changes on the ecosystem we now are broken too.

Some assitence would be great as this is getting close to 17th deadline date :/ @haroldrandom  @jim-minter

File: src/azure-cli/azure/cli/command_modules/aro/_rbac.py [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+
+    client.role_assignments.create(vnet, role_uuid, RoleAssignmentCreateParameters(
+        role_definition_id=role_definition_id,
+        principal_id=object_id,
+        principal_type='ServicePrincipal',
+    ))
+
+
+def rp_mode_development():
+    return os.environ.get('RP_MODE', '').lower() == 'development'
```
Comment (jim-minter) [inline src/azure-cli/azure/cli/command_modules/aro/_rbac.py:51]: can be removed

File: src/azure-cli/azure/cli/command_modules/aro/commands.py [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+        operations_tmpl='azure.mgmt.redhatopenshift.operations#OpenShiftClustersOperations.{}',  # pylint: disable=line-too-long
+        client_factory=cf_aro)
+
+    with self.command_group('aro', aro_sdk, client_factory=cf_aro) as g:
+        g.custom_command('create', 'aro_create', supports_no_wait=True)
+        g.custom_command('delete', 'aro_delete', supports_no_wait=True, confirmation=True)
+        g.custom_command('list', 'aro_list', table_transformer=aro_list_table_format)
+        g.custom_show_command('show', 'aro_show', table_transformer=aro_show_table_format)
+        g.custom_command('update', 'aro_update', supports_no_wait=True)
+        g.wait_command('wait')
```
Comment (jim-minter) [inline src/azure-cli/azure/cli/command_modules/aro/commands.py:24]: please confirm this works, also retrofit to ARO-RP

File: src/azure-cli/azure/cli/command_modules/aro/custom.py [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+def aro_show(client, resource_group_name, resource_name):
+    return client.get(resource_group_name, resource_name)
+
+
+def aro_list_credentials(client, resource_group_name, resource_name):
+    return client.list_credentials(resource_group_name, resource_name)
+
+
+def aro_update(client, resource_group_name, resource_name, location=None, no_wait=False):
+    oc = v2020_04_30.OpenShiftCluster(
```
Comment (jim-minter) [inline src/azure-cli/azure/cli/command_modules/aro/custom.py:137]: this is OpenShiftClusterUpdate upstream - why does it change here?  Using OpenShiftCluster here is almost certainly wrong

File: src/azure-cli/azure/cli/command_modules/aro/_rbac.py [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+from azure.cli.core.commands.client_factory import get_subscription_id
+from azure.cli.core.profiles import get_sdk
+from azure.cli.core.profiles import ResourceType
+from msrestazure.tools import resource_id
+
+
+CONTRIBUTOR = 'b24988ac-6180-42a0-ab88-20f7382dd24c'
+
+
+def assign_contributor_to_vnet(cli_ctx, vnet, object_id, label, tags):
```
Comment (jim-minter) [inline src/azure-cli/azure/cli/command_modules/aro/_rbac.py:19]: why is tags added here?

File: src/azure-cli/azure/cli/command_modules/aro/_help.py [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+
+helps['aro list-credentials'] = """
+    type: command
+    short-summary: List credentials of a cluster.
+"""
+
+helps['aro wait'] = """
+type: command
+short-summary: Wait for a aro service to reach a desired state.
+long-summary: If an operation on a aro cluster was interrupted or was started with `--no-wait`, use this command to wait for it to complete.
```
Comment (jim-minter) [inline src/azure-cli/azure/cli/command_modules/aro/_help.py:47]:
```suggestion
    type: command
    short-summary: Wait for a cluster to reach a desired state.
    long-summary: If an operation on a cluster was interrupted or was started with `--no-wait`, use this command to wait for it to complete.
```

File: src/azure-cli/azure/cli/command_modules/aro/_validators.py [hunk 16]
```diff
@@ -0,0 +1,219 @@
...
+        i = int(i)
+    except ValueError:
+        raise CLIError("Invalid --%s '%s'." % (key.replace('_', '-'), i))
+
+    return i
+
+
+def validate_pull_secret(namespace):
+    if namespace.pull_secret is None:
+        logger.warning("No --pull-secret provided: cluster will not include Red Hat or certified samples or operators.")
```
Comment (jim-minter) [inline src/azure-cli/azure/cli/command_modules/aro/_validators.py:86]: this has changed upstream and needs to be retrofitted here

File: src/azure-cli/azure/cli/command_modules/aro/commands.py (hunk 12 above)
Comment (mjudeikis) [reply src/azure-cli/azure/cli/command_modules/aro/commands.py:24]: it does.

File: src/azure-cli/azure/cli/command_modules/aro/_rbac.py [hunk 17]
```diff
@@ -0,0 +1,51 @@
...
+
+    for assignment in list(client.role_assignments.list_for_scope(vnet)):
+        if assignment.role_definition_id.lower() == role_definition_id.lower() and \
+                assignment.principal_id.lower() == object_id.lower():
+            return
+
+    if tags is None:
+        tags = {}
+
+    role_uuid = uuid.uuid5(uuid.NAMESPACE_URL, ','.join(cli_ctx.data['safe_params'] + [label, cli_ctx.data['command']] + list(tags.keys()) + list(tags.values())))  # pylint: disable=line-too-long
```
Comment (jim-minter) [inline src/azure-cli/azure/cli/command_modules/aro/_rbac.py:41]:
```suggestion
    role_uuid = uuid.uuid5(uuid.NAMESPACE_URL, ','.join(vnet, object_id, 'Contributor'))
```

**Comment by jim-minter** [comment]
@haroldrandom I approve, please merge this.
